Refactoring log functions and adding rotating log files

miniche opened the pull request:
TODO:
- [x] Refactor CentralReport log functions
- [x] Add RotatingFileHandler methods to CentralReport logs
- [x] Refactor CentralReport debug mode (if possible)

miniche committed:
[WIP] Working on rotating log in CentralReport

miniche committed:
[WIP] Working on new log functions : round 2!

miniche committed:
Working on issue #34. Updated default log file in bash scripts (now located in /tmp)

miniche committed:
Little work tonight on logs: max size, logging level and some refactors.

miniche committed:
Improved log module

miniche commented:
Ok guys, you can review this PR :grin:

NB: The refactor of debug mode and convention name will be done in next pull request.

Guimove commented:
Tested and Approved, good luck with merge conflicts

Ninir commented on bash/log.inc.sh:
```diff
@@ -7,8 +7,8 @@
-# Logs will be written in this file. Installer or Uninstaller have administrative privileges, we can write in /var/log.
-ERROR_FILE="/var/log/centralreport.log"
+# Logs will be written in this file, it doesn't require administrative privileges.
+ERROR_FILE="/tmp/centralreport_install.log"
```
I would add it in `/tmp/centralreport/`

miniche replied:
I'm not sure if it's very useful. This file is temporary, and it used only if an error occurs during install/uninstall.
And this file will be deleted immediately after a restart.

But, you give me an idea: I think we can delete this file if no error occured during installation, no?

Ninir replied:
Well, creating the folder IS useful as it will contains our log files. I think our structure should be as modular as possible, and as silent as possible, meaning we should not polluate /tmp directly, putting everything in a folder we control.

The purpose of /tmp is to be temporary, so adding a temporary folder isn't a real matter, is it? :)

miniche replied:
But we have only one file to write into /tmp: the install log.
All other logs generated by the main application go into /var/log. It isn't useless to create a folder for only tmp file?

Ninir replied:
In fact, after thinking of it again, a dir seems useless here yes... this is a `/tmp` dir, not the `/var/log` as you said.
Forget about this

miniche replied:
Oh! We agree together! Wonderful! :smile:

Ninir commented on centralreport/cr/log.py:
```diff
@@ -2,50 +2,93 @@
-        logging.basicConfig(filename='/var/log/centralreport.log',
-                            format='%(levelname)s \t %(asctime)s \t %(message)s',
-                            level=logging.INFO, datefmt='%m/%d/%Y %I:%M:%S')
+        # Max size per log file: 5 MB (1024 * 1024 * 5). Number of log files will be kept for archive: 2.
```
It seems you have something missing in your second sentence:
`Number of log files "that" will be kept for archive: 2.` ?

Btw, I would write: `2 files will be kept as archive.`

Ninir commented on centralreport/cr/log.py:
```diff
@@ -2,50 +2,93 @@
 def writeDebug(text):
-    logging.debug(text)
+    """
+        Writes a debug message. Only useful for testing and debugging purposes.
```
You could use it on your own, meaning that saying "only" is a bit restrictive.
Why not saying "Useful for testing and debugging purposes."?

miniche replied:
Yes, it's restrictive. In fact, debug message are ignored by default, not written in the log file neither on standard output.
When you launch CR with the "test.py" script, "debug mode" is enable, and these messages will be kept (log file + stdout). So, for me, "only" is the good work. I will add a clearer explanation in the comment :)

Ninir replied:
If debug messages are ignored, then "only" is well chosen, otherwise it should not be there. Thanks for sharing :)

Ninir commented on centralreport/cr/log.py:
```diff
@@ -2,50 +2,93 @@
 def writeError(text):
-    logging.error(text)
+    """
+        Writes an error message: an anormal situation, but non-critical.
```
Typo: `aBnormal`.

`Writes an error message for non-critical abnormal situations.`

Ninir commented on centralreport/cr/log.py:
```diff
@@ -2,50 +2,93 @@
 def writeCritical(text):
-    logging.critical(text)
+    """
+        Writes a critical error. This error requires to stop the current action.
```
`Writes a critical error stopping the current action.`

miniche replied:
You change the meaning of my sentence. This function doesn't stop the current action.
By "this error requires to stop the current action.", I just want to define a critical error.

Ninir replied:
Your second sentence is not crystal-clear. I know that, at the time I was writing my comment that without looking at the code, and  seeing at the doc that your function will write the log. If we translate your sentence: "Cette erreur force à stopper l'action courante."
IMO, your second sentence is simply useless as it brings no value to the function. It is only an information for you, but as a developer... I don't really care :/

miniche replied:
Ok, I understand what you want to say.
For me, it's important to define what is a critical, a fatal error, or a "simple" error, and when these functions must be used (and when we shouldn't use them). 
I will try to find a better sentence.

Ninir replied:
For instance, this is taken from the Monolog library, which is now a reference in the PHP world:

``` php
/**
* Adds a log record at the DEBUG level.
*
* @param  string  $message The log message
* @param  array   $context The log context
* @return Boolean Whether the record has been processed
*/
public function addDebug($message, array $context = array())
{
    return $this->addRecord(self::DEBUG, $message, $context);
}
```

Ninir replied:
Even better: https://github.com/Seldaek/monolog/blob/master/src/Monolog/Logger.php

Ninir commented on centralreport/cr/log.py:
```diff
@@ -2,50 +2,93 @@
 def writeFatal(text):
-    logging.fatal(text)
+    """
+        Writes a fatal error. This error requires to stop current application.
```
`Writes a fatal error stopping the application.`

miniche replied:
Same answer as the previous comment.

Ninir replied:
Same answer as my previous comment :)

miniche replied:
Same answer as my previous comment. +3

Ninir commented:
Excepted comments, seems ok. You'll have to merge conflicts... enjoy :D

miniche committed:
Add some explanations in comments (log module)

Ninir commented on centralreport/cr/log.py:
```diff
@@ -2,51 +2,94 @@
+import logging.handlers
 import sys
 
+# Our custom logger object. Initialized on the first use.
```
As you point out the fact that initialization will be done on first use in the `get_cr_logger` description, your second sentence seems useless or, at least, repetitive.

Ninir commented on centralreport/cr/log.py:
```diff
@@ -2,51 +2,94 @@
+                                                                           backupCount=2)
 
-    else:
+        custom_rotating_fileHandler.setFormatter(logging.Formatter(fmt='%(levelname)s \t %(asctime)s \t %(message)s',
```
What do you think of this format?

`[01:01:2000 01:01:01] - LEVELNAME: Message`

miniche replied:
[01:01:2000 01:01:01] - LEVELNAME \t Message

Guimove committed:
Refactor following your comments on naming convention

Ninir commented:
Here is a proposal for the logger object:

Logger Object: Here to log things at a given log level.

User functions:
- log_debug:

``` python
def log_debug(text):
    """
        Adds a log record at the DEBUG level.
    """
    get_cr_logger().debug(text)
```
- log_info:

``` python
def log_info(text):
    """
        Adds a log record at the INFO level.
    """
    get_cr_logger().info(text)
```
- log_error:

``` python
def log_error(text):
    """
        Adds a log record at the ERROR level.
    """
    get_cr_logger().error(text)
```
- log_crit:

``` python
def log_crit(text):
    """
        Adds a log record at the CRITICAL level.
    """
    get_cr_logger().critical(text)
```
- log_fatal:

``` python
def log_fatal(text):
    """
        Adds a log record at the FATAL level.
    """
    get_cr_logger().fatal(text)
```

Ninir committed:
Improved logger

- Added the warning function
- Removed fatal as duplicated of the critical function
- Added a private formatLogger function
